### ISSUE TITLE
TUN interface correctly created on the UE but internet not working

### ISSUE THREAD

**Issue by syedmiftahur** [Orange-OpenSource/towards5gs-helm#6]
Be default ipv4.ip_forward is disabled on the UPF POD. 

But when i try to set it to "1" it fails with the following output "READ ONLY FILE"
![image](https://user-images.githubusercontent.com/69756652/134487500-eb2637b4-59c2-4f56-b18f-e677a75fb571.png)

In this case any suggestion on how to overcome the problem.
As of now "**ping -I uesimtun0 www.google.com**" from the UE pod is failing

**Comment by syedmiftahur** [comment]
Although my ping is still not working through the tunnel, In the meantime I have been able to set the **ipv.ip_forward=1** in the UPF container.

Used the following procedure
sudo nsenter -t <container pid> -n sysctl -w net.ipv4.ip_forward=1

Container pid was extracted using **systemd-cgls**
From the tree output it was easy to identify the UPF - refer to the following output. 
**_│ ├─3663988 /free5gc/free5gc-upfd/free5gc-upfd -f /free5gc/config/upfcfg.yaml_**

Took tcpdump in the UPF pod - in parallel executed ping command in the UE pod.
Only the following keeps repeating. ( 10.1.0.7 is the TUN interface uesimtun0)
04:41:16.669332 IP (tos 0x0, ttl 64, id 61698, offset 0, flags [DF], proto ICMP (1), length 84)
    10.1.0.7 > maa03s40-in-f4.1e100.net: ICMP echo request, id 105, seq 4, length 64

My K8s cluster is using calico v3.20.1 - but there are no policies applied to restrict traffic outflow (refer to following screenshot)
**Do we need to make any DNS entries for  "maa03s40-in-f4.1e100.net" - Or any other suggestion is welcome.**
![image](https://user-images.githubusercontent.com/69756652/134622606-7d26199e-9f33-42aa-b347-dd95ec87d30e.png)

**Comment by raoufkh** [comment]
Hello @syedmiftahur 

IP forwarding cannot be enabed within the Kubernetes POD. Instead, you have to enable it by configuring the CNI plugin you're using. As you're using calico, you've to customize its configuration. Please see this [section](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet).

**Comment by wojt3ek140** [comment]
Hello @raoufkh

I didn't want to open a new issue as my problem is kind of similar. The ping from the UE to google is failing. I'm using the Flannel CNI plugin with default settings https://raw.githubusercontent.com/coreos/flannel/master/Documentation/kube-flannel.yml. All pods are created correctly as well as network-attachment-definitions. 

```
root@v3:/ueransim/build# traceroute -i uesimtun0 8.8.8.8
traceroute to 8.8.8.8 (8.8.8.8), 30 hops max, 60 byte packets
 1  192.168.0.216 (192.168.0.216)  6.923 ms  6.857 ms  6.827 ms
 2  10.73.4.12 (10.73.4.12)  3076.562 ms !H  3076.555 ms !H  3076.529 ms !H
```
```
Name:         n6network
Namespace:    free5gc
Labels:       app.kubernetes.io/managed-by=Helm
Annotations:  meta.helm.sh/release-name: v3.0.5
              meta.helm.sh/release-namespace: free5gc
API Version:  k8s.cni.cncf.io/v1
Kind:         NetworkAttachmentDefinition
Spec:
  Config:  { "cniVersion": "0.3.1", "type": "macvlan", "master": "eth1", "mode": "bridge", "ipam": { "type": "static", "subnet": "10.73.4.0/24", "routes": [ { "dst": "10.73.4.0/24", "gw": "10.73.4.1" } ], "gateway": "10.73.4.1" } }
Events:    <none>
```

UPF has the correct (I think) interface also with IP forwarding enabled:
```
6: n6@if3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 4a:04:97:49:da:15 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.73.4.12/24 brd 10.73.4.255 scope global n6

root@v3:/free5gc/free5gc-upfd# cat /proc/sys/net/ipv4/ip_forward
1

root@v3:/free5gc/free5gc-upfd# ip route
default via 192.168.0.1 dev eth0 
10.1.0.0/17 dev upfgtp proto static 
10.73.4.0/24 dev n6 proto kernel scope link src 10.73.4.12 
10.100.50.232/29 dev n3 proto kernel scope link src 10.100.50.233 
10.100.50.240/29 dev n4 proto kernel scope link src 10.100.50.241 
10.244.0.0/16 via 192.168.0.1 dev eth0 
192.168.0.0/24 dev eth0 proto kernel scope link src 192.168.0.216 
```

During the pod creation, I changed the values.yaml files in order to match IP subnets and gateways to my eth interfaces:
```
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.73.4.38  netmask 255.255.255.0  broadcast 10.73.4.255
 
eth1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.73.4.195  netmask 255.255.255.0  broadcast 10.73.4.255
```

It seems like packets are reaching the UPF but it has problems with the forwarding. I will really appreciate it if you can take a look at this problem and give me some hints on how to resolve it.

**Comment by raoufkh** [comment]
> But such results are visible even when trying to traceroute local interface
This is due to MACVLAN limitations. A MACVLAN interface cannot communicate with its master interface. Howerver, it should be able to communicate with other interfaces which are on the same subnet. Have you another VM in your environment which has an interface on this subnet? If not you may try to reach eth0 interface as it is on the same subnet as eth1 and UPF N6 and it is not the master interface of N6.

**Comment by wojt3ek140** [comment]
I turned on the "Forged transmits" setting in the networking tab in the ESXi and it did the trick :)

Thank you so much for your time and help @raoufkh!

**Comment by raoufkh** [comment]
Good news! Thank you too for sharing your experience with this deployment based on vmware infrastructure!

**Comment by anhdangkhoa** [comment]
> Hello @syedmiftahur
> 
> IP forwarding cannot be enabed within the Kubernetes POD. Instead, you have to enable it by configuring the CNI plugin you're using. As you're using calico, you've to customize its configuration. Please see this [section](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet) and this [link](https://docs.projectcalico.org/reference/host-endpoints/forwarded)

Hello @raoufkh 

I have followed the instruction in this [section](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet) to configure Calico.

However, the `ipv4.ip_forward` is still disabled in UPF POD as i perform `cat /proc/sys/net/ipv4/ip_forward` and gets `0`.

Am I missing something?
I have remove the deployment and reinstall free5gc after reconfiguration.

**Comment by zanattabruno** [comment]
> 



> > Hello @syedmiftahur
> > IP forwarding cannot be enabed within the Kubernetes POD. Instead, you have to enable it by configuring the CNI plugin you're using. As you're using calico, you've to customize its configuration. Please see this [section](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet) and this [link](https://docs.projectcalico.org/reference/host-endpoints/forwarded)
> 
> Hello @raoufkh
> 
> I have followed the instruction in this [section](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet) to configure Calico.
> 
> However, the `ipv4.ip_forward` is still disabled in UPF POD as i perform `cat /proc/sys/net/ipv4/ip_forward` and gets `0`.
> 
> Am I missing something? I have removed the deployment and reinstalled free5gc after reconfiguration.

I have same issue. The ping from uesimtun0 to UPF n6if works, but no work to ping n6 gw.
And ipv4.ip_forward is still disabled in UPF POD.

**Comment by zanattabruno** [comment]
> > 
> 
> > > Hello @syedmiftahur
> > > IP forwarding cannot be enabed within the Kubernetes POD. Instead, you have to enable it by configuring the CNI plugin you're using. As you're using calico, you've to customize its configuration. Please see this [section](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet) and this [link](https://docs.projectcalico.org/reference/host-endpoints/forwarded)
> > 
> > 
> > Hello @raoufkh
> > I have followed the instruction in this [section](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet) to configure Calico.
> > However, the `ipv4.ip_forward` is still disabled in UPF POD as i perform `cat /proc/sys/net/ipv4/ip_forward` and gets `0`.
> > Am I missing something? I have removed the deployment and reinstalled free5gc after reconfiguration.
> > I have same issue. The ping from uesimtun0 to UPF n6if works, but no work to ping n6 gw.
> > I apply calico config, enable promiscuos mode and forged transmits in hypervisor too.
> > I ping if n6 from uesimtun0
> > And ipv4.ip_forward is still disabled in UPF POD.

In my case, the problem was in the Kubernetes cluster. Just apply [calico config](https://projectcalico.docs.tigera.io/reference/host-endpoints/forwarded) not work for me (Detail, I made my cluster with [kubespray](https://github.com/kubernetes-sigs/kubespray)).
I reset my cluster and enabled the [following variable](https://github.com/kubernetes-sigs/kubespray/blob/master/roles/network_plugin/calico/defaults/main.yml#:~:text=calico_allow_ip_forwarding%3A%20false) before re-deploying.
